### PR TITLE
plugins: add new plugin for Synaptics CAPE devices

### PR DESCRIPTION
**Pull request by CNXT-Simon**
CAPE family is Audio DSP for a board range of applications in IOT, PC and mobile  can be interfaced via I2C, UART or USB
interface. This patch is only for CX31993 and CX31988 chips, there is not
immediate plans to add support to other CAPE devices.

CX31993 have two separate firmware .hid file for for each partition. It
need to convert two .hid files into a .fw file for fwupd tool to consume.

Currently, this patch is only support for EPOS headsets with basic
firmware update feature. Either new code singing or manifest.xml are
unsupported yet.

The code has been tested with CX31993 EVK board.

A test firmware file is put at 'src/fuzzing/firmware/synaptics-cape.fw'

Type of pull request:

- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

# Checklist 

- [x] Fill out `README.md` with update protocol
- [x] Fill out `README.md` with any custom quirks and flags
- [x] Fill out `README.md` with the vendor ID security value
- [x] Implement `FuFirmware->write()` and include at least one fuzzer testcase in `src/fuzzing/firmware` for any custom `FuFirmware` subclass
- [x] CI run of the plugin for at least one target
- [ ] Document targets CI isn't currently run and the reasons (i.e. minimum versions needed or distribution limitations etc).

### CODE REVIEW

**Comment by hughsie** [comment]
Looks good. Let me polish up a tiny bit and then port to the new ABI.

**Comment by CNXT-Simon** [comment]
Sorry, I found errors. stay tuned for new update.

Simon

**Comment by hughsie** [comment]
@CNXT-Simon just do a fixup commit on top of mine, don't force push the branch yet please.

**Comment by hughsie** [comment]
@CNXT-Simon we can squash this down to just one commit when we're ready to merge.

**Comment by CNXT-Simon** [comment]
*   @CNXT-Simon<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_CNXT-2DSimon&d=DwMCaQ&c=7dfBJ8cXbWjhc0BhImu8wQ&r=OFh8QRRfgkCyTxu3ETj_SihTxtA9DrwQheChfFVLfSE&m=g_ufzuUPeaWO9ckbjjzIqlsyKyhAAba4bDWguoA3dn8&s=QIzBO1TKwKvFzAxx290WyEswZhhba3gBaK43LM8_1iE&e=> just do a fixup commit on top of mine, don't force push the branch yet please.

Sure, will do.

Simon

**Comment by hughsie** [comment]
@CNXT-Simon I think you have to add a line to contrib/ci/oss-fuzz.py to actually get the fuzzer to run on your new file.

File: plugins/synaptics-cape/fu-synaptics-cape-firmware.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+fu_synaptics_cape_firmware_get_vid(FuSynapticsCapeFirmware *self);
+
+guint16
+fu_synaptics_cape_firmware_get_pid(FuSynapticsCapeFirmware *self);
+
+#define FW_CAPE_HID_HEADER_OFFSET_VID	      0x0
+#define FW_CAPE_HID_HEADER_OFFSET_PID	      0x4
+#define FW_CAPE_HID_HEADER_OFFSET_UPDATE_TYPE 0x8
+#define FW_CAPE_HID_HEADER_OFFSET_SIGNATURE   0xc
+#define FW_CAPE_HID_HEADER_OFFSET_CRC	      0x10
```
Comment (hughsie) [inline plugins/synaptics-cape/fu-synaptics-cape-firmware.h:33]: should fwupd be checking the CRC too?

**Comment by CNXT-Simon** [comment]
> @CNXT-Simon we can squash this down to just one commit when we're ready to merge.

I don't know how to squash this down to just one commit using PR approach. Could you help on this? or show me how to make it.

File: plugins/synaptics-cape/fu-synaptics-cape-firmware.h (hunk 1 above)
Comment (CNXT-Simon) [reply plugins/synaptics-cape/fu-synaptics-cape-firmware.h:32]: No, the CRC is done at firmware side.

**Comment by CNXT-Simon** [comment]
> The daemon should give you the actual numbers to use. The reason I wasn't convinced was that your plugin doesn't implement ->attach() and so it seems unlike it takes 2% of the total time.

Here is log contains timestamp for each function, could you suggest how do I update the number correctly.

`
15:48:34:0776 FuDeviceList         no replug or re-enumerate required
15:48:34:0776 FuEngine             prepare -> FuSynapticsCapeDevice:
15:48:34:0948 FuPlugin             prepare(logind)
15:48:34:0952 FuPluginLogind       opened logind fd 18
15:48:34:0952 FuDeviceList         no replug or re-enumerate required
15:48:34:0952 FuDeviceList         no replug or re-enumerate required
15:48:34:0952 FuDeviceList         no replug or re-enumerate required
15:48:34:0952 FuEngine             detach -> FuSynapticsCapeDevice:
15:48:34:0952 FuPlugin             running superclassed detach(synaptics_cape)
15:48:35:0129 FuDeviceList         no replug or re-enumerate required
15:48:35:0129 FuEngine             update -> FuSynapticsCapeDevice:
15:48:35:0129 FuPlugin             superclassed write_firmware(synaptics_cape)
15:48:35:0261 FuDevice             installing onto 48cdf390f0a1768e3fb0bcb10af8fcfb38e5d89e:
15:48:45:0612 FuPluginSynapticsCape reset took 3002.06ms
15:48:45:0612 FuProgress           raw timing data was { 0.159, 7.177, 0.013, 3.002 } -- steps were set as [ 1 91 3 5 ] but should have been [ 2 69 0 29 ] at ../plugins/synaptics-cape/fu-synaptics-cape-device.c:592
15:48:45:0612 FuHidDevice          ignoring: USB error on device 1395:0280 : No such device (it may have been disconnected) [-4]
15:48:45:0612 FuDeviceList         waiting 10000ms for replug
15:48:45:0617 FuEngine             auto-removing backend device EPOS ADAPT 160 ANC [48cdf390f0a1768e3fb0bcb10af8fcfb38e5d89e]
15:48:45:0617 FuDeviceList         waiting 10000ms for EPOS ADAPT 160 ANC device removal
15:48:45:0635 FuDevice             using 48cdf390f0a1768e3fb0bcb10af8fcfb38e5d89e for usb:08:00:01
15:48:45:0635 FuHidDevice          autodetected HID interface of 0x03
15:48:45:0776 FuPlugin             emit added from synaptics_cape: 48cdf390f0a1768e3fb0bcb10af8fcfb38e5d89e
15:48:45:0783 FuPlugin             fu_plugin_device_registered(dell_dock)
15:48:45:0784 FuPlugin             fu_plugin_device_registered(thunderbolt)
15:48:45:0784 FuPlugin             fu_plugin_device_registered(pci_bcr)
15:48:45:0784 FuPlugin             fu_plugin_device_registered(msr)
15:48:45:0784 FuDeviceList         found existing device 48cdf390f0a1768e3fb0bcb10af8fcfb38e5d89e
15:48:45:0784 FuDeviceList         ::changed 48cdf390f0a1768e3fb0bcb10af8fcfb38e5d89e
15:48:45:0784 XbSilo               ignoring for OR statement: indexed string 'c481ba5a-4b2f-5f42-b351-76193e49703b' was unfound
15:48:45:0784 XbSilo               ignoring for OR statement: indexed string '8114d066-2ccc-52f6-b299-cd4d47997108' was unfound
15:48:45:0784 XbSilo               ignoring for OR statement: indexed string 'c481ba5a-4b2f-5f42-b351-76193e49703b' was unfound
15:48:45:0784 XbSilo               ignoring for OR statement: indexed string '8114d066-2ccc-52f6-b299-cd4d47997108' was unfound
15:48:45:0919 FuDeviceList         waited for replug
15:48:45:0919 FuEngine             attach -> FuSynapticsCapeDevice:
15:48:45:0919 FuPlugin             running superclassed attach(synaptics_cape)
15:48:46:0113 FuDeviceList         no replug or re-enumerate required
15:48:46:0113 FuEngine             reload -> FuSynapticsCapeDevice:
15:48:46:0293 FuDeviceList         no replug or re-enumerate required
15:48:46:0293 FuDeviceList         no replug or re-enumerate required
15:48:46:0293 FuEngine             cleanup -> FuSynapticsCapeDevice:
15:48:46:0293 FuDeviceList         no replug or re-enumerate required
15:48:46:0293 FuDeviceList         no replug or re-enumerate required
15:48:46:0293 FuEngine             cleanup -> FuSynapticsCapeDevice:
`

**Comment by hughsie** [comment]
Lines like this should help you:

    15:48:45:0612 FuProgress raw timing data was { 0.159, 7.177, 0.013, 3.002 } -- steps were set as [ 1 91 3 5 ] but should have been [ 2 69 0 29 ] at ../plugins/synaptics-cape/fu-synaptics-cape-device.c:592

Is there not an additional one further down in the log for the attach process?

**Comment by CNXT-Simon** [comment]
> Lines like this should help you:
> 
> ```
> 15:48:45:0612 FuProgress raw timing data was { 0.159, 7.177, 0.013, 3.002 } -- steps were set as [ 1 91 3 5 ] but should have been [ 2 69 0 29 ] at ../plugins/synaptics-cape/fu-synaptics-cape-device.c:592
> ```

Thanks, the message "--steps were set as [...] but should have bee [...]" is gone after I updated suggested number [2 69 0 29]. 
 
> Is there not an additional one further down in the log for the attach process?

No, I doesn't see additional one for attach process. Please see attached log file for details.
[fwupd_log.txt](https://github.com/fwupd/fwupd/files/7164079/fwupd_log.txt)

**Comment by hughsie** [comment]
Ahh, add back the _GUESSED in the fu_synaptics_cape_device_set_progress() function, reinstall then try again. If it's "close enough" then we don't show the output.

File: plugins/synaptics-cape/fu-synaptics-cape-device.c [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+}
+
+static void
+fu_synaptics_cape_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+
+	g_return_if_fail(FU_IS_SYNAPTICS_CAPE_DEVICE(self));
+
+	fu_common_string_append_ku(str, idt, "active_partition", self->active_partition);
```
Comment (hughsie) [inline plugins/synaptics-cape/fu-synaptics-cape-device.c:292]: `fu_common_string_append_ku(str, idt, "ActivePartition", self->active_partition);` please

**Comment by hughsie** [comment]
One little nit to fix, and then please do `git rebase master -i` squash/fixup all the commits into one commit and then `git push --force` to update this PR with the beautiful new plugin. Thanks.

**Comment by CNXT-Simon** [comment]
> That looks more like I was expecting, thanks!

The timing self-checking check is cool!

File: plugins/synaptics-cape/fu-synaptics-cape-device.c (hunk 2 above)
Comment (CNXT-Simon) [reply plugins/synaptics-cape/fu-synaptics-cape-device.c:292]: oops, I don't know why this is back. I am working on it.

Comment (hughsie) [reply plugins/synaptics-cape/fu-synaptics-cape-device.c:292]: That might have been one of my over-zealous search/replaces. If so, apologies.

**Review by hughsie — APPROVED** [review]

**Pull request merged**